params/include/alps/params/dict_value_impl.hpp

```diff
@@ -0,0 +1,574 @@
+#include <boost/type_traits/is_signed.hpp>
+#include <boost/type_traits/is_unsigned.hpp>
+#include <boost/type_traits/is_floating_point.hpp>
+#include <boost/type_traits/make_unsigned.hpp>
+#include <boost/type_traits/make_signed.hpp>
+#include <boost/integer_traits.hpp>
+#include <boost/type_index.hpp> // for pretty-printing exceptions
+
+#include <alps/utilities/short_print.hpp> // for streaming of vectors
+
+namespace std {
+template <typename T>
+inline std::ostream& operator<<(std::ostream& s, const std::pair<std::string,T>& p) {
+    s << "{" << p.first << "," << p.second << "}";
+    return s;
+}
+
+
+template <typename T>
+inline std::ostream& operator<<(std::ostream& s, const std::vector<T>& v) {
+    s << alps::short_print(v);
+    return s;
+}
+}
+
+namespace alps {
+    namespace params_ns {
+
+
+        namespace detail {
+            template <bool V> struct yes_no {};
+            template <> struct yes_no<true> : public boost::true_type { typedef bool yes; };
+            template <> struct yes_no<false> : public boost::false_type { typedef bool no; };
+
+            template <typename T>
+            struct is_bool : public yes_no<boost::is_same<T,bool>::value> {};
+
+            // bool type is NOT integral for the purposes of this code
+            template <typename T>
+            struct is_integral : public yes_no<boost::is_integral<T>::value && !is_bool<T>::value> {};
+
+            // signed value: integral and signed
+            template <typename T>
+            struct is_signed : public yes_no<boost::is_signed<T>::value && is_integral<T>::value> {};
+
+            // unsigned value: integral and unsigned
+            template <typename T>
+            struct is_unsigned : public yes_no<boost::is_unsigned<T>::value && is_integral<T>::value> {};
+
+            // meta-predicate: conversion bool->integral
+            template <typename FROM, typename TO>
+            struct is_bool_to_integral
+                : public yes_no<is_bool<FROM>::value && is_integral<TO>::value>
+            {};
+
+            // meta-predicate: conversion signed->integral
+            template <typename FROM, typename TO>
+            struct is_sig_to_intgl
+                : public yes_no<is_signed<FROM>::value && is_integral<TO>::value>
+            {};
+
+            // meta-predicate: conversion unsigned->integral
+            template <typename FROM, typename TO>
+            struct is_unsig_to_intgl
+                : public yes_no<is_unsigned<FROM>::value && is_integral<TO>::value>
+            {};
+
+            // meta-predicate: conversion integral->floating_point
+            template <typename FROM, typename TO>
+            struct is_intgl_to_fp
+                : public yes_no<is_integral<FROM>::value && boost::is_floating_point<TO>::value>
+            {};
+
+            // meta-predicate: general conversion, not caught by other ones
+            template <typename FROM, typename TO>
+            struct is_other_conversion
```

Comment:
Do you need a special predicate to detect *the rest* of cases?
I'd say it's compiler's duty to find the default overload of `apply()`.

Reply:
I'd expect it to be the compiler's duty, but it somehow did not work for me. I may be missing something here.

Reply:
I suspect these overloads are not mutually exclusive. Something fishy is going on with those predicates.
Code compiles well (no ambiguous overloads reported) if I comment out all but two functions.
```c++
                    /// Simplest case: the values are of the same type
                    LHS_T apply(const LHS_T& val) const {
                        return val; // no conversion
                    }

                    /// Placeholder: extracting any other type
                    template <typename RHS_T>
                    LHS_T apply(const RHS_T& val) const {
                        std::string rhs_name=boost::typeindex::type_id<RHS_T>().pretty_name();
                        std::string lhs_name=boost::typeindex::type_id<LHS_T>().pretty_name();
                        throw exception::type_mismatch("","Types do not match;"
                                                       " conversion " + rhs_name + " --> " + lhs_name);
                    }
```